### ISSUE TITLE
Snowflake - fun with fractals

### ISSUE THREAD

**Issue by RobinSchmidt** [RobinSchmidt/RS-MET#194]
i'm working on a new module that can create waveshapes that trace out a closed in 2D with a self-similar boundary. here's an example:

![image](https://user-images.githubusercontent.com/1833099/36623960-33331cca-190a-11e8-83b1-e98723214e13.png)

here are the x- and y-coordinates as function of time:

![image](https://user-images.githubusercontent.com/1833099/36623976-47001442-190a-11e8-96a7-8371a8520593.png)

as can be seen, the waveforms are periodic, i.e. they end where they begin, indicating a nicely loopable repetitive waveform. someone posted on facebook a link to the gif from here:

https://www.reddit.com/r/visualizedmath/comments/7xtxgb/hilbert_curve/

which got me interested in what the fuck is going on there. a bit of research revealed that they are drawing a Moore curve there:

https://en.wikipedia.org/wiki/Moore_curve

and doing some more research was dragging me into a rabbit hole. the moore curve is one of the infinitely many curve types that a Lindenmayer-system:

https://de.wikipedia.org/wiki/Lindenmayer-System

will be able to generate. so i implemented one. among the possible curves is also the koch snowflake:

 https://en.wikipedia.org/wiki/Koch_snowflake

which i elected for the name of this new upcoming generator.

**Comment by RobinSchmidt** [comment]
here are more examples:

http://mathforum.org/advanced/robertd/lsys2d.html

not all of the curves shown there loop back to their beginning though, but with some care in setting up the L-system, this can be ensured (it's a useful property when it's going to be used as audio oscillator waveform). here's a little web-app to play around with L-systems until my new module is ready:

http://www.kevs3d.co.uk/dev/lsystems/

**Comment by RobinSchmidt** [comment]
i have no idea, how it will sound like, but i'm pretty sure, it will make for great visuals on the scope :-)

**Comment by elanhickler** [comment]
The stereo 1D image version of the waveform looks like a supersaw. If you can do subtle modulation of the details of the waveform this could make for an interesting chorusy waveform.

**Comment by RobinSchmidt** [comment]
i currently can't see any straightforward way to modulate the synthesis (i.e. rendering) parameters as these parameters are either strings or an integer (the number of iterations) and also, re-rendering of the (wave) tables is computationally expensive as hell. the only continuous parameter is the angle, but it doesn't seem to make much sense to modulate it - it must be set to a very specific value in accordance to chosen rules in order to get a closed loop.

but maybe some post-processing can be slapped on the waveform. maybe an allpass with different frequency for left and right channel. i think, that would shift the overtones a bit back and forth

**Comment by RobinSchmidt** [comment]
...but perhaps it could make sense to do a kind of crossfading between various similar pre-rendered wavetables as in waldorf sense of wavetable synthesis

**Comment by RobinSchmidt** [comment]
> the only continuous parameter is the angle, but it doesn't seem to make much sense to modulate it

thinking about it some more, i think modulating the turning angle in realtime (if that would be computationally possible in any meaningful way) would give results similar to just modulating a post-processed 2D rotation of the resulting picture. maybe not quite but similar

**Comment by RobinSchmidt** [comment]
> perhaps it could make sense to do a kind of crossfading between various similar pre-rendered wavetables

..when i integrate this generator into my upcoming dream-synth with its 4 sources that are crossfadable via a vector pad, i could set all 4 sources to similar L-system shapes and modulate the x/y coordinates of the vector mixer pad

**Comment by elanhickler** [comment]
Make it modulatable, I can't imagine it is more expensive than my synths Radar, and Spiral Generator.

**Comment by RobinSchmidt** [comment]
hmmm...i really need to think about it. maybe it's possible to make the turning angle of the L-system modulatable. ...because it's actually not a parameter of the L-system itself but of the "turtle graphics" interpreter. i think, the turtle can be modified to produce one point at a time (instead of rendering the whole table of points at once). i have an idea how that may be possible....but first things first

**Comment by RobinSchmidt** [comment]
first audio signals produced. 
![image](https://user-images.githubusercontent.com/1833099/36634557-9db56ec6-19a6-11e8-9f07-1af43843656f.png)
maybe in a few hours or tomorrow, it's ready to actually play with it and explore the range of possibilities (some important basics like state-recall are still missing). first impression of the sound: it has a sort of organ-like quality to it with more iterations having the effect of adding in more upper registers or something - which is actually not really surprising at all, when thinking about it

**Comment by RobinSchmidt** [comment]
i just found a link that i need to save:
http://algorithmicbotany.org/papers/#abop
i feel like having stumbled upon a whole new branch of mathematics that has nothing to do with what i have seen before (apart from some introduction into formal languages in computer science lectures during study). ...but all the "normal" math stuff - geometry, linear algebra, calculus, complex numbers - even number theory - all that seems to be irrelevant here - but somehow it's still math. math is sooo diverse. amazing! but maybe it's not *mathematics* but *algorithmics*? but that latter term seems to be not be widely used

**Comment by RobinSchmidt** [comment]
even the fractals that i have encountered so far were based on a totally different mechanism. in the most general case, you had a vector valued function y = f(x) where x and y are vectors and f is some (nonlinear) function that maps and input vector onto an output vector...and then you apply that function again and again and again...that makes a nonlinear recursive system. and for some systems, the iteration diverges for some initial vectors and doesn't diverge for others. the boundary between those classes of starting points has a fractal shape. the well known mandelbrot set:

https://en.wikipedia.org/wiki/Mandelbrot_set

is a famous example of that. ...but this here is conceptually completely different

**Comment by RobinSchmidt** [comment]
> the complex-number formulation of the function f(z) can straightforwardly be translated into a function mapping 2D vectors onto 2D vectors. i think:

`f(z) = z^2 + c`

translates to:

```
x_new = x^2 - y^2 + c
y_new = 2 * x * y + c 
```
when you want to do it without complex numbers. iterate that and observe, if the vector explodes or not. paint starting-points for which it doesn't explode black and points where is does explode white and you get a basic (black/white) picture of the mandelbrot set. artistic renderings often take into account, how fast it diverges and choose a color according to that value.

**Comment by RobinSchmidt** [comment]
> maybe it's not mathematics but algorithmics?

...but why does it feel so totally different? i think, maybe it's because one of the most fundamental concepts of math  - the equation - does not really play a big role there?

**Comment by RobinSchmidt** [comment]
by the way - i just noticed that my text editor seems to behave buggy as hell. does it work for you as it should in the comment section of chaosfly?

**Comment by RobinSchmidt** [comment]
ooookaaaay....the fun can begin! there are still a couple of quirks and bugs but at least it's now possible to explore L-systems. i created a new repository with presets and i already created some quite interesting shapes here:

https://github.com/RobinSchmidt/Presets/tree/master/Snowflake

very interesting is to load MooreCurve4, put an EngineersFilter after that, set it to bessel lowpass (of somewhat high order) and tweak the cutoff frequency. i that reminds you of the gif of the rolling circles that initially sparked my interest in all this stuff, it should. it's a sort of replication of what's going on there (just with a bessel filter instead of a perfect linear-phase brickwall filter which would correspond to the gif). very satisfying. ok...tomorrow maybe a bit more exploration and playtime and then back to work on the multicomp gui. but i just *had* to do this

**Comment by RobinSchmidt** [comment]
btw. - can you recommend me a software for doing screen capture videos? i may want to make some in the future as well....

**Comment by RobinSchmidt** [comment]
endless variation of interesting shapes, not necessarily self-similar
![image](https://user-images.githubusercontent.com/1833099/36650455-dfd32430-1aa3-11e8-85d5-816932ef290c.png)

**Comment by RobinSchmidt** [comment]
damn! you are right! i *have* to make the turning angle modulatable. this is not like just rotating the whole picture. it's actually much more interesting. if you want to try it, i recommend to load the patch PentaT (it's the patch that created the pic above) and move the "Angle" slider by hand. of course, there are lots of artifacts now because changing the angle triggers a re-rendering. but if i implement it in a way that expects the angle to be modulated, i think, that could make for a great effect - both sonically and visually.

**Comment by elanhickler** [comment]
> by the way - i just noticed that my text editor seems to behave buggy as hell. does it work for you as it should in the comment section of chaosfly?

It has been buggy from the beginning.

> btw. - can you recommend me a software for doing screen capture videos? i may want to make some in the future as well....

if you have a good or modern video card: https://obsproject.com/ (you have to set video card capturing option if you have a modern video card, nvidia GTX or radeon RX to take advantage of video card based capture).

if you have an old or crappy video card: https://getsharex.com/

**Comment by RobinSchmidt** [comment]
i have many ideas, how to extend the turtle-graphics and lindenmayer-system syntax in meaningful ways to allow for even more variation of the geometry. i think, a whole new synthesis paradigm can be created from this...i'm very excited about this stuff (as you have undoubtetly noticed). i think, i'll call it Fractal Geometry (FG) synthesis. ..it needs a catchy name

**Comment by RobinSchmidt** [comment]
an evil alien from independence day :-O
![image](https://user-images.githubusercontent.com/1833099/36796601-327d52ce-1ca6-11e8-94cc-54772b246319.png)
.. and this patch also sounds good (AlienFace.xml)

**Comment by elanhickler** [comment]
pooooost auudioooooooooooo

looks great.

**Comment by RobinSchmidt** [comment]
i do not yet have the means to produce video with audio - but will hopefully do soon. you can try it yourself. just build chainer, plug in Snowflake into the first slot (and trigger a (preferably low) note) and skip through the presets (just download the preset repo into standard preset folder C:\Users\[Username]\AppData\Roaming\RS-MET\Presets). maybe plug a BreakpointModulator into another slot and route it multiplicatively to the amplitude for basic amp-enveloping (but it behaves sometimes a bit buggy).

...i want to fix a couple of bugs and then also get back to the multicomp gui. ...but it's really a rabbithole. i'm just scratching the surface. using filters on the output is also sometimes very useful - especially for the moore curve. if you want to preserve the shape best, use bessel filters

**Comment by RobinSchmidt** [comment]
...and i really need to write up a tutorial at some point to explain what (the fuck) is going on there, so a sound designer can actually know what (the fuck) he's doing. it's very unlike common synthesis methods

**Comment by RobinSchmidt** [comment]
> can i sell any of this as a soundemote product? what do you want to do? do you want to work together at all?

i want to use it as one of the generators in my upcoming dream synth - but yeah, sure, if you want to also make a synth from it for your oscilloscope-synth series, we can certainly work together on that

> I can make you big bucks. My company will grow.

that would be great!

**Comment by elanhickler** [comment]
ok, it is good to know you are open to the idea. Nothing has to happen right away, I will let my mind work on this.

**Comment by RobinSchmidt** [comment]
i just fixed a bug - the was a sort of buzzing artifact when CyclicReset was 0. it didn't sound necessarily bad, but it was wrong anyway. this is fixed now. the MooreRotor had it before - it's gone now

**Comment by elanhickler** [comment]
Ok this is pretty cool. I tried it. You need phase and frequency modulation! You also need LFOs and my feedback system.

You need to...... just copy what I'm doing with Jerobeam stuff 😄 

Or I can take your core module and put it inside an OMS product and give you the same (or similar royalties) deal I gave Jerobeam.

**Comment by elanhickler** [comment]
I will have suggestions before I make it a Soundemote product.

1. Add a control for "the length of the pattern" from 0 to 100%, but the frequency should remain the same. Essentially, moving this control will give you a "slowly being drawn" effect on an oscilloscope.

2. Can you think of a way, without a filter, to add/remove harmonic content? Like some modulation of some part of the algorithm that happens to make the shape less spikey, more rounded. If not, then make a dedicated non-resonant filter for snowflake.

**Comment by RobinSchmidt** [comment]
> Or I can take your core module and put it inside an OMS product and give you the same (or similar royalties) deal I gave Jerobeam.

yes - that sounds good. i think, you have the surrounding environment already in place and would just have to plug in the core sound generator at the center. what is your arrangement with jerobeam?

> Add a control for "the length of the pattern" from 0 to 100%, but the frequency should remain the same. Essentially, moving this control will give you a "slowly being drawn" effect on an oscilloscope.

how would that work? are you talking about traversing the pattern only partially and then reset? "slowly being drawn" is something i would just associate with playing a low frequency but i guess i'm misunderstanding you. here

> Can you think of a way, without a filter, to add/remove harmonic content? Like some modulation of some part of the algorithm that happens to make the shape less spikey, more rounded. If not, then make a dedicated non-resonant filter for snowflake.

yes, just a few hours ago i was think about that. for many patterns, the number of iterations adds harmonic content (well, for some you get totally different patterns, but that may also be due some mistakes in defining is - i'm just learning all that myself). ...sooo a continuous number-of-iterations would be nice. when it's set to 3.5, you'd get a 50/50 crossfade between 3 and 4 iterations. well, this is my idea. but a non-resonant bessel filter can be plugged after it anyway. also, regular resonant filters (like my ladder) are also nice. you loose some of the symmetries that bessel filters would preserve but it often looks good anyway

**Comment by RobinSchmidt** [comment]
> some modulation of some part of the algorithm that happens to make the shape less spikey, more rounded.

i also want to include cubic modulation later. it currently works like that: the drawing algorithm (the "turtle") generates a sequence of x,y points, between which i linearly interpolate to generate the time signal. using cubic interpolation instead would also round edges. ...and one could also crossfade beween linear and cubic interpolation. and/or left/right/nearest neighbour interpolation. there, the signal would jump from spot to spot, so on the scope only the corner points would be visible and in the time-signal, we would have discontinuous jumps - with all their harmonic creating properties

**Comment by RobinSchmidt** [comment]
this DragonCurve.xml here looks and sounds quite good:
![image](https://user-images.githubusercontent.com/1833099/36862932-95bfb22c-1d88-11e8-8cdc-15762c8d1d77.png)
also, the number of iterations increases harmonic content in a useful way.
https://en.wikipedia.org/wiki/Dragon_curve  (i'm using a closed loop version of it)

**Comment by elanhickler** [comment]
very nice. You frequency is too high for that image to be drawn accurately though. :)

**Comment by RobinSchmidt** [comment]
indeed! here's a better rendering using a very low frequency and different colormap:
![image](https://user-images.githubusercontent.com/1833099/36870183-929107d4-1d9e-11e8-8815-e8270888b3ec.png)
haha! this is really fun! btw. the number of iterations is already maxed out at 10 here in this pic. it could actually use one or two more levels of detail

**Comment by elanhickler** [comment]
Are you doing or can you do anything to lessen harmonics as frequency increases (as an antialiasing technique)?

**Comment by RobinSchmidt** [comment]
OMG! FUCK YEAH!!! i just added a feature for resetting the turtle after a given number of lines - and this is the KILLER! pull an update and check out the new patches AlienFace2 and 3 and BrokenTriangles2. sounds and looks already amazing. now tweak the TurningAngle parameter!

**Comment by RobinSchmidt** [comment]
i actually think, that the resetting strategy should be considered as another major aspect of the synthesis method. just resetting after completing a given number of cycles through the curve and (additionally, independently) resetting after some line counter has reached its maximum is very simple. i think, i should come up with more sophisticated resetting algorithms. i'm thinking of using something like my line-count based resetter, but not reset every time, the number of lines is reached, but let that depend on the value of some binary sequence (1 means: do reset, 0 means: don't reset) ...and then look into interesting ways to generate such sequences. one possibility would be to use L-systems once again and interpret '+' as '1' and '-' as '0'. but there are other ways as well. on monday, i received my copy of this wonderful book:

http://www.abrazol.com/books/patterngen/

...for 14 euros at amazon, that was a no brainer - and it's a gold-mine. containing lots of pictures with construction information.  the AlienFace patch is actually one of those. ...and - besides L-systems - it also talks about other methods to generate binary sequences. i may implement some of them as well in the future

**Comment by RobinSchmidt** [comment]
ah - the examples are online is we - it's actually this:

http://www.abrazol.com/books/patterngen/image/trees60d04.svg

from here:

http://www.abrazol.com/books/patterngen/images_lsys.html

...but my rendering looks different...hmm...have i messed up something? i need to figure that out. if so, i think, it would count as happy accident

**Comment by elanhickler** [comment]
have an option to reverse direction instead of just reset. Although I could add that myself with phase modulation. You have a phase variable?

**Comment by RobinSchmidt** [comment]
yes - that's a good idea. i'll have to see how hard or easy that will turn out to implement. i do have a "phase" variable (it's `pos` in class `TurtleSource`)...but there's no easy and straightforward way to just "set" it in a random access way. to set it to some arbitrary value, i would actually have to traverse the curve from the current position to the desired target position - which may be expensive. ...sooo...i guess...phase-modulation is difficult (or i must think harder)...but frequency modulation seems to be no big deal, i think

**Comment by elanhickler** [comment]
thennnn forget about phase modulation!

then make a phase mod to frequency mod converter... haha... I was going to pay you for this when I can.

**Comment by RobinSchmidt** [comment]
...unless the optional table-mode is used (where i just pre-render the whole curve into a table and read from that) - in this case, randomly setting the position is trivial. but that mode is limited in other ways

**Comment by elanhickler** [comment]
don't do table mode.

**Comment by RobinSchmidt** [comment]
that's why it's optional. it's actual a remnant of the first implementation - it was just simpler to do. but i think i'll don't throw the code away but keep it as option. it's also useful for debugging - writing unit tests and compare outputs from table mode and non-table mode. when they don't match, there's a bug somewhere

**Comment by elanhickler** [comment]
you can have negative frequency, right? wouldn't that reverse the direction? so that's part of the function of converting phase mod to freq mod.

**Comment by RobinSchmidt** [comment]
not yet - but yes, this is the way i'm planning to implement to reverse direction. my turtle does not yet support to go backward, nor does my realtime turtle-driver support to go backward in the turtle command string. but i hope that this will be straightforward to add

**Comment by RobinSchmidt** [comment]
> i think, i should come up with more sophisticated resetting algorithms

...aaand i just discovered that the same authors that wrote the pattern generation book wrote another one which may be just what i need:

http://abrazol.com/books/rhythm1/

i think, we may need resetting algo-rhythms! instantly ordered. :-) ...and actually they wrote some more:

http://www.exstrom.com/stefan/stefan.html

some of which might be interesting later. especially the one about finite automata

**Comment by RobinSchmidt** [comment]
hmmm...i just discovered a problem. i implemented the possibility to render only a partial picture by resetting the drawing algorithm early, before it could finish drawing the whole picture. ok - straightforward enough and works as expected. next i wanted to let the user adjust the start-position inside the drawing, i.e. that start "phase" to make it possible to draw different parts. now there's a problem: it works only for some patches, namely those which don't use saving and restoring the turtle state during drawing (if you have no idea what that means, look at this document (still very much under construction): https://github.com/RobinSchmidt/RS-MET/blob/master/Documentation/Tutorials/FractalPatternSynthesis.md)

hmmm...that's sort of bad - although, when trying the feature in patches where it does work, it didn't seem to be that useful - at least from a sonic point of view, so i disabled it for now (by just making the slider invisible).  i would expect the same (or a similar) problem to appear when i introduce support negative frequencies / time-reversed read-out

**Comment by RobinSchmidt** [comment]
sooo...i need to think, about ways to reconcile start-position modulation and time-reversing with the turtle state stack. in case of reverse readout, i think, it could be done by swapping the interpretations of `[` and `]` during reverse readout. i think, i must also swap left-turns for right-turns and vice versa and of course taking a step backward instead of forward on `F`

**Comment by RobinSchmidt** [comment]
ok - i made some good progress. it supports negative frequencies now (mostly - with some L-system definitions, it goes out of whack) and also phase-modulation. skipping through the presets is already good fun. there some real gems in there already.

i'll put it on the shelf for the next week for work on the multicomp gui and maybe do some preset design on the side for recreation

**Comment by RobinSchmidt** [comment]
i have now also implemented the feature for periodic reversal of the direction. seems to work. there are now two independent resetters and one reverser - each of which has its own periodicity/interval which acts independently from the others. theses intervals can be set up as a multiple/fraction of the signal period. ...plus some frequency dependent additive offset - the idea is that when the different resetters/reversers have slightly different intervals, we hear a sort of modulation. using these offsets, we can keep the "modulation-frequency" constant across the keyboard.

the remaining missing feature: continuous adjustment of the iteration number is more complicated. i'm working on the multicomp gui now, so this shall be postponed until i'm done with this. ...ah and also anti-aliasing will have to be added later (although, one could always simply do an oversampling wrapper - but i hope, i can do better)

**Comment by elanhickler** [comment]
Alright, I think my release schedule will be this:

1. Update all Oscilloscope Music synths
2. Release free version of Filters of Mass Desctruction
3. Release Fractal Generator
4. Release paid version of FMD

**Comment by RobinSchmidt** [comment]
3. Release Fractal Generator (pay you in royalties)

did you already wrap the snowflake code into your framework? maybe, we should talk about the exact conditions first. 

> my arrangement with Jerobeam is 30% royalties.
> So I can give you 30% royalties unless you are responsible for more of the synth (if you basically create the final product mostly or completely on your own then I would give you more). But consider that I will be doing marketing, GUI design, including all my personal DSP, etc.

hmm...well - i am providing not just the algorithm prototypes but the ready-to-use c++ code along with a whole bunch of (pre)-presets. i'd actually consider a 50/50 share more fair (consider that i have also written the framework under your framework)...but then, i don't really care that much because if you make lots of money with it, you have more funds to hire me for other work...so whatever...

**Comment by elanhickler** [comment]
My responsibilities include:
- website management
- advertisement, promotion, social media
- audio demos
- video demos
- creating the gui (using your framework)
- testing
- adding in my own DSP (somewhat built on your DSP, but some of my code is my own, AND some of the code I I paid you for already)
- artist for art assets for gui (planned)

Your responsibilities
- research
- DSP and programming
- testing
- framework

If you want 50% then I'll ask you to take on more responsibility.

**Comment by RobinSchmidt** [comment]
maybe we should talk about this on a non-public channel. i'll send an email soon

**Comment by RobinSchmidt** [comment]
i just stumbled over this today:
http://benice-equation.blogspot.de/2012/01/fractal-spirograph.html
i wonder, what kind of equation is behind that. it's definitely something very different from my turtle/lindenmayer approach (it seems to be based on revolving cirlces). ..i .need to explore...might make for another fractal synthesis algo based on a totally different idea (and still not even being close to the mandelbrot approach)

**Comment by elanhickler** [comment]
crashes at 0 frequency! plz fix!

**Comment by elanhickler** [comment]
![](https://i.imgur.com/0OlSIzF.png)

![](https://i.imgur.com/1N1png8.png)

![](https://i.imgur.com/TznaTN7.png)

![](https://i.imgur.com/b1tCVVu.png)

![](https://i.imgur.com/lz7ZBdX.png)

![](https://i.imgur.com/RFbLfZj.png)

![](https://i.imgur.com/QoSdAqD.png)

**Comment by BHMOO028** [comment]
I believe you have found me. Can you please insure that is real and not a redo to trick me. And please update my github. Thank you very much.

Get Outlook for Android

________________________________

**Comment by RobinSchmidt** [comment]
> crashes at 0 frequency! plz fix!

should be fixed

> Edit: Antialiasing button seems to do nothing

yep - as stated above - it's not yet implemented

cool pix!

**Comment by RobinSchmidt** [comment]
btw. - hilbert curves are actually useful:
https://www.youtube.com/watch?v=3s7h2MHQtxc&index=2&list=PLZHQObOWTQDO__zBYmoxntqx3yBpuXQBl

**Comment by RobinSchmidt** [comment]
> I believe you have found me. Can you please insure that is real and not a redo to trick me. And please update my github. 

?? i don't understand that. who are you?

**Comment by elanhickler** [comment]
probably a spam post.

**Comment by RobinSchmidt** [comment]
prettyscope and snowflake are a great combo for creating art!

**Comment by RobinSchmidt** [comment]
i made a new filter prototype that should be ideally suited for artifact-free parameter modulation and can realize (almost) any biquad transfer function (there are singular cases where is can only approximate it with great precision - but that shouldn't matter in practice). it will probably be useful for implementing a modulatable bessel filter. ...but now really back to the cycle marks

**Comment by elanhickler** [comment]
I don't get it, what do you mean artifact-free parameter modulation? Parameter modulation artifacts depend on the audio signal being outputted, not by the modulation. Or do you mean this is only useful for the modulatable bessel filter? Or do other filters also benefit? Confused as to how I might benefit. 😄

**Comment by RobinSchmidt** [comment]
think of a regular biquad filter, straightforwardly implemted in what is called "direct from 1" in dsp lingo. the output sample y[n] at sample instant n is computed as:
y[n] = b0*x[n] + b1*x[n-1] + b2*x[n-2] - a1*y[n-1] - a2*y[n-2]
i.e. as  a weighted sum of the current inputs sample x[n], two previous input samples x[n-1], x[n-2] and two previous output samples y[n-1], y[n-2]. the weighting coefficients b0, b1, b2, a1, a2 are computed from the user parameters (cutoff, reso, filter-type, etc.) - the formulas for these computations is where all the complicated dsp-math is located. now, the user changes the cutoff and or resonace or whatever other parameter. coefficients have to be recomputed. now, the stored "past values" (y[n-1], etc.) sort of become invalid. consider the case of setting the cutoff from 2kHz to 500Hz - from one sample to the next. the new coefficients together with the old filter state may create a sort of transient in the output signal

**Comment by RobinSchmidt** [comment]
there's a recent thread in the kvr dsp forum about exactly this issue:
http://www.kvraudio.com/forum/viewtopic.php?f=33&t=506213
now, the challenge is to come up with other filter structures (different from this "direct form 1") that exhibits better responses to sudden changes of the coefficients (using internal state variables other than the past in/out samples). i think, i have found a structure that should be well behaved in this respect. ...some experimentation testing is needed, but i think, the lions share of the work is done (at least as far as i can see from a theoretical point of view). my "filter-state" is a 2D vector and the recursive part of the filter scales and rotates this vector (rotation angle depends on cutoff/resonance frequency, decay depends on Q) - and suddenly changing these things will not affect the validity of the state vector

**Comment by RobinSchmidt** [comment]
`do other filters also benefit?`
yes - any filter that can be expressed as a biquad (or serial or parallel connection of biquads) can benefit (parallel would actually be better). ...which means pretty much any filter. engineers filter is a series connection of biquads. with the current implementation, i wouldn't even think of trying to modulate a high-order elliptic filter's cutoff with a square wave - it would probably completely mess up. with the new structure, it might be practical (albeit expensive - but i'm just talking about stability and transients here)

**Comment by elanhickler** [comment]
I wonder if that means you'd want to use one of your engineer's filters to do FFT-like operations where modulating the filter cutoff is required to track, for example, the fundamental and other partials of an audio.

**Comment by RobinSchmidt** [comment]
indeed, it could be useful for partial extraction when the frequency of the partial is changing and the filter must track these changes. but in partial extraction, i would probably want to apply it bidirectionally and i currently have no idea if such a bidirectional filter will preserve its desirable zero-phase property when being modulated. for this reason, maybe a time-variant FIR might be still better in this case. but it's something that may be worth to explore. but it was not really the main goal here

**Comment by RobinSchmidt** [comment]
> my "filter-state" is a 2D vector and the recursive part of the filter scales and rotates this vector (rotation angle depends on cutoff/resonance frequency, decay depends on Q)

in this respect, it is actually the same thing that i did with my PhasorFilter which is conceptually more or less the same as this:

https://ccrma.stanford.edu/courses/250a-fall-2003/hiqfilters.pdf

 the new thing is that the new filter does not necessarily couple the x- and y-coordinates in a spiraling decay - they can now also decay exponentially independently - when being mixed together, this corresponds to two first order lowpasses in parallel that are added (with mixing coefficients). this extension makes almost every biquad representable by this structure. the only problem is a filter that has two identical (real) poles. but it can be approximated very closely by spreading the poles a little

**Comment by elanhickler** [comment]
Where are you at with this? I want to use Snowflake to generate some funds so I can pay you royalties and hourly.
